src/libcore/pin.rs

```diff
@@ -94,6 +95,106 @@
+//! to pinned data. For example, in an intrusive doubly-linked list, every element
+//! will have pointers to its predecessor and successor in the list. Every element
+//! will be pinned, because moving the elements around would invalidate the pointers.
+//! Moreover, the `Drop` implemenetation of a linked list element will patch the pointers
+//! of its predecessor and successor to remove itself from the list. Clearly, if an element
+//! could be deallocated or overwritten without calling `drop`, the pointers into it
+//! from its neighbouring elements would become invalid, breaking the data structure.
+//!
+//! Notice that this guarantee does *not* mean that memory does not leak! It is still
+//! completely okay not to ever call `drop` on a pinned element (e.g., you can still
+//! call [`mem::forget`] on a `Pin<Box<T>>`). What you may not do is free or reuse the storage
+//! without calling `drop`.
+//!
+//! # `Drop` implementation
+//!
+//! If your type relies on pinning (for example, because it contains internal
+//! references, or because you are implementing something like the intrusive
+//! doubly-linked list mentioned in the previous section), you have to be careful
+//! when implementing `Drop`: notice that `drop` takes `&mut self`, but this
+//! will be called even if your type was previously pinned! It is as if the
+//! compiler automatically called `get_unchecked_mut`. This can never cause
+//! a problem in safe code because implementing a type that relies on pinning
+//! requires unsafe code, but be aware that deciding to make use of pinning
+//! in your type (for example by implementing some operation on `Pin<&[mut] Self>`)
+//! has consequences for your `Drop` implemenetation as well.
+//!
+//! # Projections and Structural Pinning
+//!
+//! One interesting question arises when considering pinning and "container types" --
+//! types such as `Vec` or `Box` but also `RefCell`; types that serve as wrappers
+//! around other types.  When can such a type have a "projection" operation, an
+//! operation with type `fn(Pin<&[mut] Container<T>>) -> Pin<&[mut] T>`?
+//! This does not just apply to generic container types, even for normal structs
+//! the question arises whether `fn(Pin<&[mut] Struct>) -> Pin<&[mut] Field>`
+//! is an operation that can be soundly added to the API.
+//!
+//! This question is closely related to the question of whether pinning is "structural":
+//! when you have pinned a container, have you pinned its contents? Adding a
+//! projection to the API answers that question with a "yes" by offering pinned access
+//! to the contents.
+//!
+//! In general, as the author of a type you get to decide whether pinning is structural, and
+//! whether projections are provided. However, there are a couple requirements to be
+//! upheld when adding projection operations:
+//!
+//! 1. The container must only be [`Unpin`] if all its fields are `Unpin`. This is the default,
```

Comment:
The requirement is only that the container is `Unpin` if all fields it projects to are `Unpin`, you can have a `!Unpin` field and be `Unpin` as long as there is no way to project to that field (in this way `Unpin` being an auto-trait is actually too conservative since most types don't provide any way to project to their fields, but so far I haven't seen any real usability issues arise from it).

Reply:
Yeah I didn't want to go into "mixed containers" where you can only project to some fields. I felt that would get too complicated. But maybe it's not too bad.